src/arch/arm/kernel/boot.c

```diff
@@ -304,7 +312,19 @@ BOOT_CODE static void release_secondary_cpus(void)
     /* Wait until all the secondary cores are done initialising */
     while (ksNumCPUs != CONFIG_MAX_NUM_NODES) {
         /* perform a memory release+acquire to get new values of ksNumCPUs */
-        __atomic_signal_fence(__ATOMIC_ACQ_REL);
+        __atomic_thread_fence(__ATOMIC_ACQ_REL);
+#if defined(CONFIG_KERNEL_MCS) && defined(CONFIG_DEBUG_BUILD)
+        /* Test whether clocks are synchronised across nodes: */
+        ticks_t t = getCurrentTime();
+        for (int i = 1; i < CONFIG_MAX_NUM_NODES; ++i) {
+            ticks_t tc = clock_sync_test[i];
+
+            if (tc != 0) {
+                assert(tc <= t && t <= tc + getTimerPrecision());
```

Comment:
I have some doubt that there is a 100% guarantee that `tc <= t`, because this sequence could happen
```
    Core 1                              Core 0
                                     t = getCurrentTime()
t1 = getCurrentTime() (= t + n)
clock_sync_test[1] = t1
                                     tc = clock_sync_test[1] (= t + n)
                                     assert, since tc is higher than t
```
This might be unlikely to happen, but can't be rules out. Should this check better allow `t` to be in the range `tc - getTimerPrecision()/2` to `tc + getTimerPrecision()/2`? Another way might be taking a snapshot, then take the time and then check it against the snapshot. The assumption here is, that a timestamp taken after making the snapshot must be greater or equal than anything in the snapshot 
```
#if defined(CONFIG_KERNEL_MCS) && defined(CONFIG_DEBUG_BUILD)
        ticks_t snapshot[CONFIG_MAX_NUM_NODES];
        for (int i = 1; i < CONFIG_MAX_NUM_NODES; ++i) {
            snapshot[i] = clock_sync_test[i];
        }
        /* ensure the snapshot is taken before we get the time below */
        __atomic_thread_fence(__ATOMIC_ACQUIRE);
        ticks_t t = getCurrentTime(); /* this cannot be less than an snapshot value */
        for (int i = 1; i < CONFIG_MAX_NUM_NODES; ++i) {
            ticks_t tc = snapshot[i];
            if (tc != 0) {
                assert(tc <= t && t <= tc + getTimerPrecision());
                clock_sync_test[i] = 0;
            }
        }
```

Reply:
Well spotted. It's a fairly theoretical problem though: As `ksNumCPUs` is only modified while holding the BLK, node 0 has time equal to the BLK hold time to do the check. But it does mean that the data ordering with `ksNumCPU`  is forced and unnecessary, so all the barriers could go away if this race is fixed.

A simpler fix would be to just always take a timestamp within the `if (tc != 0)`. I'll have a look at improving the code.

Reply:
> A simpler fix would be to just always take a timestamp within the if (tc != 0). I'll have a look at improving the code.

I have also considered. The downside of this would be, that the timestamp is "drifting" away with each call, but practically that delta might be negligible if `getTimerPrecision()`is  large enough. On the other side it has to be large enough to cope with the situation that we miss an update of `clock_sync_test`, reach the end of the loop and then start over again, which calls `getCurrentTime()` again then anyway. Bottom line would be that we assume calling `getCurrentTime()` is fast and cheap, so it does no matter how often we do it.

Reply:
Okay, I think I missed that I could expand the individual test results.